File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
const solution = (numbers, target) => {
    
    let result = 0;
    
    const dfs = (start, depth, n, visited) => {
        //타겟넘버 계산
        if(depth === n) {
            let targetNumber = 0;
            for(const element of numbers) {
                targetNumber += element;
            }
        
            if(targetNumber === target) result++;
            
            return;
        }
        
        for(let i = start; i < numbers.length; i++) {
            if(!visited[i]) {
                visited[i] = true;
                numbers[i] *= -1; 
                
                dfs(i+1, depth + 1, n, visited);
                
                visited[i] = false;
                numbers[i] *= -1;
            }
        }
        
        return;
    }
    
    for(let i = 0; i < numbers.length; i++) {
        let visited = Array.from({length: numbers.length}, () => false);
        dfs(0, 0, i+1, visited);
    }
    
    return result;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.io.*;
import java.util.*;

public class Solution {
    static int answer;
    
    public static int solution(int[] numbers, int target) throws IOException {
        
        answer = 0;
        
        dfs(0, target, numbers);
        
        return answer;
    }
    
    static void dfs(int idx, int target, int[] arr) {
        if(idx >= arr.length) {
            int tmp = 0;
            for(int i = 0; i < arr.length; i++) {
                tmp += arr[i];
            }
            
            if(tmp == target) {
                
                answer += 1;
                return;
            }
            return;
        }
        
        
        dfs(idx + 1, target, arr);
        
        arr[idx] *= -1;
        
        dfs(idx + 1, target, arr);
    }
}